### ISSUE TITLE
Have Autokey shortcut activate only when in a text field

### ISSUE THREAD

**Issue by osbornejr** [autokey/autokey#266]
## Classification:

 Enhancement

## Reproducibility:

Always

## Version

AutoKey version: 0.95.1 

Used GUI (Gtk, Qt, or both): Gtk

Installed via: PPA

Linux Distribution: Ubuntu 18.04

## Summary

I am trying to change the start/end line input keys in Ubuntu from `home` and `end` to `ctrl`+`right` and `ctrl`+`left`. This is very simple to do in Autokey using phrases and works like a charm in Atom and in text fields in my web browser. The problem is that the shortcut is overridden at all times i.e. I cannot use `ctrl`+`right` or `ctrl`+`left` for anything else now... I use those for going back and forward in my browser, as well as switching between applications. 

Is it possible to set the phrase so that it works system-wide, but only when in a text field (i.e. a blinking cursor)?. I imagine that would occur via the window filter, but that seems to only allow activation per application.

**Comment by josephj11** [comment]
You're right about how window filters work. I don't know of a way to do this.

For questions (as opposed to issues), it's better to post on our [support list](https://groups.google.com/forum/#!forum/autokey-users) or on [Gitter](https://gitter.im/autokey/autokey).
A lot more people will see it and may be able to help.

Gitter would be better for this question because that's where the developers and advanced users hang out.

**Comment by osbornejr** [comment]
OK thank you, I will head over there to see if anyone has any ideas.

**Comment by Elliria** [comment]
I realize this was filed four years ago, but I figured I'd jump in with a possible way to get this done.

You could write an AutoKey conditional script that checks if you're currently in a text field (the method of testing for this will depend on the program the field is in, so I leave that part up to the reader. If you're not currently in a text field, then pass. If you're currently in a text field, then running one of the scripts below should work as the action with a bit of modification based on what your abbreviation is, whether you're using a trigger, what kind of trigger you're using, etc. We can help with that, though.

These abbreviations were used in the examples with various settings for the typed abbreviation and trigger character:
* The single-character examples use **x**.
* The multiple-duplicate-character examples use **xxx**.
* The multiple-non-duplicate-character example uses **xyz**.
* The single- or multiple-character example uses **xyz** followed by a press of the space bar.

Run a phrase from a script if the phrase has a single-character abbreviation and no trigger:
```
keyboard.fake_keypress("x")
```
or:
```
keyboard.press_key("x")
keyboard.release_key("x")
```

Run a phrase from a script if the phrase has a multiple-duplicate-character abbreviation and no trigger:
```
keyboard.fake_keypress("x", repeat=3)
```
or:
```
keyboard.press_key("x")
keyboard.release_key("x")
keyboard.press_key("x")
keyboard.release_key("x")
keyboard.press_key("x")
keyboard.release_key("x")
```

Run a phrase from a script if the phrase has a multiple-non-duplicate-character abbreviation and no trigger:
```
keyboard.fake_keypress("x")
keyboard.fake_keypress("y")
keyboard.fake_keypress("z")
```

Run a phrase from a script if the phrase has a single- or multiple-character abbreviation and a trigger:
```
keyboard.fake_keypress("x")
keyboard.fake_keypress("y")
keyboard.fake_keypress("z")
keyboard.fake_keypress(" ")
```
***Note:** The `keyboard.send_key()` and `keyboard.send_keys()` API calls will not work for this.*

**Comment by josephj11** [comment]
@Elliria I don't understand what you are proposing. Is this something you have actually tested?

AFAIK, nothing in AutoKey rescans the output of AutoKey for triggers. If fake keypresses get around this, that's a big deal.

The engine API includes run_script, but AFAIK, it does not have anything like run_phrase. Convoluting this, one could use the `<script>` macro in a phrase, but I've never understood what this was good for that was simpler or better than just using a script instead of a phrase.

As far as detecting whether the text/input cursor is in a text field, I'm also at a loss. While it is possible to get the current relative or absolute mouse cursor position, AFAIK, there is no similar facility for the text/input cursor. I don't think that visgrep() would help with this either.

**Comment by Elliria** [comment]
Yes, this is something I've tested, but not fully. I didn't figure out the code for the OP's condition, which is to establish whether the active cursor is currently in a text field. That will be up to the OP and will depend on how (each of) the program(s) lets you determine the environment the cursor is currently in.

I did, however, run a phrase from a script in several ways, as shown above. You can follow along with this basic demonstration of one of the ways that uses a Boolean condition:
1. Create a phrase.
2. Make these its contents: bar
3. Make this its abbreviation: xxx
4. Enable **Remove typed abbreviation** in the abbreviation settings.
5. Enable **Trigger when typed as part of a word** in the abbreviation settings.
6. Enable **Trigger immediately (don't require a trigger character)** in the abbreviation settings.
7. Save the phrase.
8. Create a script.
9. Pick one of these as its contents:
   ```
   foo = True
   if foo == True:
     keyboard.press_key("x")
     keyboard.release_key("x")
     keyboard.press_key("x")
     keyboard.release_key("x")
     keyboard.press_key("x")
     keyboard.release_key("x")
   else:
     pass
   ```
   or:
   ```
   foo = True
   if foo == True:
     keyboard.fake_keypress("x", repeat=3)
   else:
     pass
   ```
   or:
   ```
   foo = True
   if foo == True:
     keyboard.fake_keypress("x")
     keyboard.fake_keypress("x")
     keyboard.fake_keypress("x")
   else:
     pass
   ```
10. Save the script.
11. Open a text file.
12. Run the script. *The result should be that AutoKey should print **bar** in the text editor.*
13. Edit the script and change **True** to **False**.
14. Save the script.
15. Open the text file.
16. Run the script again. *The result should be that nothing should happen.*

**Comment by Elliria** [comment]
> Convoluting this, one could use the `<script>` macro in a phrase, but I've never understood what this was good for that was simpler or better than just using a script instead of a phrase.

According to the [Dynamic Phrases, Using Macros as placeholders in Phrases](https://github.com/autokey/autokey/wiki/Dynamic-Phrases,-Using-Macros-as-placeholders-in-Phrases) page (we really need to rename that), there are several rules and no real examples. A hint is given, though, in that although we can use a script to alter background system-state, like by starting or stopping system services, we'd get the best results by restricting ourselves to reading data in.

I imagine that means that if you had a phrase that doesn't change and some ever-changing data that you'd like it to contain, then a script could be used to read the data in its current state. Like you said, though, the same could be accomplished with a script. I guess it's all about options, since everyone has preferences for how to get things done.

**Comment by josephj11** [comment]
The first case needs to be in the wiki. I didn't think that was possible. What happens if you leave out steps 5 and or 6? Does it still trigger on a non-word character or not work at all? I'd like to understand why it works too.

Dynamic phrases: using special purpose macros in phrases ??

I think the main thing is that  the macro is replaced by its final value (or action - in the case of `<cursor>`) and attempting to output directly from within the macro either does nothing or breaks things.

**Comment by Elliria** [comment]
> The first case needs to be in the wiki.

The single-character abbreviation one?

> I didn't think that was possible.

Me, neither. It took quite a while before I figured out to even try it.

> What happens if you leave out steps 5 and or 6? Does it still trigger on a non-word character or not work at all?

The same thing that would happen if you did it with any abbreviation that you abbreviated/triggered manually. Nothing would happen if it's part of a word and/or nothing would happen unless you followed it by immediately pressing a non-word key, like the space bar. You could get around the second one by adding a press of the space-bar to the script.

> I'd like to understand why it works too.

It works because AutoKey doesn't care how keys are pressed, but whether they're pressed. Since it happens to be capable of pressing keys, it can fulfill its own obligation with a script. It's interesting that `keyboard.send_key()` and `keyboard.send_keys()` don't work, though, but I suppose that's because they're technically the equivalent of Python's `print` command rather than a key-press.

> Dynamic phrases: using special purpose macros in phrases ??

Maybe. It's certainly shorter.

> I think the main thing is that the macro is replaced by its final value (or action - in the case of `<cursor>`) and attempting to output directly from within the macro either does nothing or breaks things.

We were specifically looking at the `<script...>` macro, which has its own set of rules. I've messed around with them a bit a long time ago and managed to get them to fall over more often than I managed to get them to succeed, but it's probably because I broke the rules in my attempts.

**Comment by josephj11** [comment]
I said that wrong. I mean the example(s) in which you actually get it to work with whatever key sequences.

**Comment by Elliria** [comment]
All of the examples above work. I just provided three different ways to do the same thing so the user can pick which one he/she prefers.

**Comment by Elliria** [comment]
I [added it to the wiki](https://github.com/autokey/autokey/wiki/Advanced-Scripts#run-an-autokey-phrase-from-an-autokey-script) on the [Advanced Scripts](https://github.com/autokey/autokey/wiki/Advanced-Scripts) page just above the one that shows how to run a script from a script.
.
